### PR TITLE
ROX-17322: ignore scrape updates

### PR DESCRIPTION
**Pull request by jschnath**
## Description

Ignore ScrapeUpdates when in Offline mode. The commandHandlerImpl component received a bool flag to determine whether central is reachable and now only sends ScrapeUpdates when it is reachable.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Added Unit tests

### CODE REVIEW

File: sensor/common/compliance/command_handler_impl.go [hunk 1]
```diff
@@ -45,7 +49,14 @@ func (c *commandHandlerImpl) Stop(_ error) {
 	c.stopper.Client().Stop()
 }
 
-func (c *commandHandlerImpl) Notify(common.SensorComponentEvent) {}
+func (c *commandHandlerImpl) Notify(e common.SensorComponentEvent) {
+	switch e {
+	case common.SensorComponentEventCentralReachable:
+		c.centralReachable.CompareAndSwap(false, true)
```
Comment (lvalerom) [inline sensor/common/compliance/command_handler_impl.go:55]: Why not `Store`?

**Review by lvalerom — COMMENTED** [review]
Partial review since I saw the unit tests were failing. I think it looks good. There are a couple of minor comments regarding when `centralReachable` should be `true`.

Also, would it make sense to avoid calling `commitResults` if we are offline?

**Review by fredrb — COMMENTED** [review]
+1 to all @lvalerom's comments above

File: sensor/common/compliance/command_handler.go [hunk 2]
```diff
@@ -17,6 +19,9 @@ type CommandHandler interface {
 
 // NewCommandHandler returns a new instance of a CommandHandler using the input image and Orchestrator.
 func NewCommandHandler(complianceService Service) CommandHandler {
+	reachable := &atomic.Bool{}
+	reachable.Store(true)
```
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler.go]: Oh yeah this one was redundant

File: sensor/common/compliance/command_handler_impl.go [hunk 3]
```diff
@@ -37,6 +40,7 @@ func (c *commandHandlerImpl) ResponsesC() <-chan *message.ExpiringMessage {
 }
 
 func (c *commandHandlerImpl) Start() error {
+	c.centralReachable.Store(true)
```
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler_impl.go:43]: For this one I disagree, when the component is started doesn't that happen through a command from central, eg. central is reachable at that moment? Or does central explicitly send another signal that it is ready after starting the components?

File: sensor/common/compliance/command_handler_impl.go (hunk 1 above)
Comment (jschnath) [reply sensor/common/compliance/command_handler_impl.go]: It doesn't really make a difference, I changed it to Store so it's easier to read.

File: sensor/common/compliance/command_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+func (s *CommandHandlerTestSuite) TestCommandHandlerStops() {
+	s.mockService.EXPECT().Output()
+	err := s.cHandler.Start()
+	s.Require().NoError(err)
+	s.True(s.cHandler.centralReachable.Load())
+	s.cHandler.Notify(common.SensorComponentEventOfflineMode)
+	s.False(s.cHandler.centralReachable.Load())
+	s.cHandler.Notify(common.SensorComponentEventCentralReachable)
+	time.Sleep(2000 * time.Millisecond)
+	s.True(s.cHandler.centralReachable.Load())
```
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler_test.go:75]: I've added tests checking for messages regarding the `commandHandler.updates()` channel

**Review by fredrb — APPROVED** [review]
It looks good now, thanks for changing the tests Jan. Before merging, check my response to Luis' comment related to Start/Stop functionality.

File: sensor/common/compliance/command_handler_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler_impl.go:43]: Okay, thanks for clearing that up, I've changed it.

File: sensor/common/compliance/command_handler.go [hunk 5]
```diff
@@ -17,6 +19,9 @@ type CommandHandler interface {
 
 // NewCommandHandler returns a new instance of a CommandHandler using the input image and Orchestrator.
 func NewCommandHandler(complianceService Service) CommandHandler {
+	reachable := &atomic.Bool{}
+	reachable.Store(false)
```
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler.go:23]: Okay, done

File: sensor/common/compliance/command_handler_impl.go [hunk 6]
```diff
@@ -25,7 +27,8 @@ type commandHandlerImpl struct {
 
 	scrapeIDToState map[string]*scrapeState
 
-	stopper concurrency.Stopper
+	stopper          concurrency.Stopper
+	centralReachable *atomic.Bool
```
[the remark this answers is not in the document]
Comment (jschnath) [reply sensor/common/compliance/command_handler_impl.go:31]: Okay

File: sensor/common/compliance/command_handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/compliance/command_handler.go:23]: It looks like the `reachable` variable is still there - we can remove it.

**Comment by vikin91** [comment]
I think you need to rebase to make the check green - retest won't help as this is not OSCI but a Github action.

File: sensor/common/compliance/command_handler.go [hunk 7]
```diff
@@ -17,6 +19,9 @@ type CommandHandler interface {
 
 // NewCommandHandler returns a new instance of a CommandHandler using the input image and Orchestrator.
 func NewCommandHandler(complianceService Service) CommandHandler {
+	reachable := &atomic.Bool{}
```
Comment (fredrb) [inline sensor/common/compliance/command_handler.go:22]: The `reachable` variable is still here. Make sure to remove it before merging.

**Comment by vikin91** [comment]
It looks like there is a data race in the unit tests:

```
=== RUN   TestCommandHandler
=== RUN   TestCommandHandler/TestCommandHandlerStops
=== RUN   TestCommandHandler/TestNoResponseWhenOffline
==================
WARNING: DATA RACE
Write at 0x00c000b55bf8 by goroutine 64:
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).SetupTest()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:32 +0x44a
  github.com/stretchr/testify/suite.Run.func1()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:187 +0x252
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c000b55bf8 by goroutine 63:
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).run()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:100 +0x6a4
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).Start.func1()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:43 +0x39

Goroutine 64 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  github.com/stretchr/testify/suite.runTests()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:242 +0x19c
  github.com/stretchr/testify/suite.Run()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:215 +0x9b2
  github.com/stackrox/rox/sensor/common/compliance.TestCommandHandler()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:19 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 63 (finished) created at:
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).Start()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:43 +0xd6
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).startCommandHandler()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:88 +0x193
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).TestCommandHandlerStops()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:96 +0x9e
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:728 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:370 +0xc7
  github.com/stretchr/testify/suite.Run.func1()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:197 +0x70d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
common/compliance: 2023/08/09 12:27:45.738349 command_handler_impl.go:140: Info: started scrape "foo" with 2 results desired
    testing.go:1446: race detected during execution of test
=== RUN   TestCommandHandler/TestResponseWhenCentralReachable
common/compliance: 2023/08/09 12:27:46.240307 command_handler_impl.go:140: Info: started scrape "bar" with 2 results desired
=== RUN   TestCommandHandler/TestStartScrapeSucceedsComplianceMessageNotSent
==================
WARNING: DATA RACE
Write at 0x00c000b55bd0 by goroutine 68:
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).SetupTest()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:32 +0x44a
  github.com/stretchr/testify/suite.Run.func1()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:187 +0x252
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c000b55bd0 by goroutine 67:
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).run()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:85 +0x184
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).Start.func1()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:43 +0x39

Goroutine 68 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  github.com/stretchr/testify/suite.runTests()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:242 +0x19c
  github.com/stretchr/testify/suite.Run()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:215 +0x9b2
  github.com/stackrox/rox/sensor/common/compliance.TestCommandHandler()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:19 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 67 (finished) created at:
  github.com/stackrox/rox/sensor/common/compliance.(*commandHandlerImpl).Start()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_impl.go:43 +0xd6
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).startCommandHandler()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:88 +0x193
  github.com/stackrox/rox/sensor/common/compliance.(*CommandHandlerTestSuite).TestResponseWhenCentralReachable()
      /__w/stackrox/stackrox/sensor/common/compliance/command_handler_test.go:120 +0xa4
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:728 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:370 +0xc7
  github.com/stretchr/testify/suite.Run.func1()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:197 +0x70d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
common/compliance: 2023/08/09 12:27:46.240860 command_handler_impl.go:140: Info: started scrape "foo" with 2 results desired
    testing.go:1446: race detected during execution of test
=== RUN   TestCommandHandler/TestStartScrapeSucceedsComplianceMessageSent
common/compliance: 2023/08/09 12:27:46.742197 command_handler_impl.go:140: Info: started scrape "foo" with 2 results desired
=== NAME  TestCommandHandler
    testing.go:1446: race detected during execution of test
--- FAIL: TestCommandHandler (3.01s)
    --- PASS: TestCommandHandler/TestCommandHandlerStops (2.00s)
    --- FAIL: TestCommandHandler/TestNoResponseWhenOffline (0.50s)
    --- PASS: TestCommandHandler/TestResponseWhenCentralReachable (0.00s)
    --- FAIL: TestCommandHandler/TestStartScrapeSucceedsComplianceMessageNotSent (0.50s)
    --- PASS: TestCommandHandler/TestStartScrapeSucceedsComplianceMessageSent (0.00s)
```

**Pull request merged**